File: Maths/List_All_Primes_2.py
"""
Write a program that takes an integer argument and returns all prime numbers between 1 and the integer (included).

Sieve of eratosthenes approach.

Time complexity: O(Nlog(logN))
"""

def listAllPrimes(n:int)->list:
	primes = []
	is_prime = [True for i in range(n+1)]
	is_prime[0] = is_prime[1] = False
	
	for p in range(2,n+1):
		if is_prime[p]:
			primes.append(p)
		# sieve multiple of p
		for j in range(p,n+1,p):
			is_prime[j]=False
	return primes
	
if __name__=="__main__":
	n = int(input())
	print(listAllPrimes(n),end=" ")

File: Maths/List_All_Primes_2.cxx
/*
 * Write a program that takes an integer argument and returns all prime numbers between 1 and the integer (included).
 * 
 * Sieve of eratosthenes approach.
 * 
 * Time complexity: O(Nlog(logN))
*/

#include <bits/stdc++.h>
using namespace std;

vector<int> allprimes(int n){
	vector<int> primes;
	vector<bool> is_prime(n+1,true);
	is_prime[0] = is_prime[1] = false;
	for(size_t p = 2; p<=n; p++){
		if(is_prime[p]){
			primes.emplace_back(p);
		}
		for(size_t j = p; j<=n; j+=p){
			is_prime[j] = false;
		}
	}
	return primes;
}

int main(){
	ios_base::sync_with_stdio(false);
	int n;
	scanf("%d",&n);
	for(size_t x:allprimes(n)){
		cout << x << " ";
	}
	return 0;
}
